libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -881,6 +915,75 @@ string YulUtilFunctions::readFromStorageDynamic(Type const& _type, bool _splitFu
 	});
 }
 
+string YulUtilFunctions::readFromMemory(Type const& _type)
+{
+	return readFromMemoryOrCalldata(_type, false);
+}
+
+string YulUtilFunctions::readFromCalldata(Type const& _type)
+{
+	return readFromMemoryOrCalldata(_type, true);
+}
+
+string YulUtilFunctions::readFromMemoryOrCalldata(Type const& _type, bool _fromCalldata)
+{
+	string functionName =
+		string("read_from_") +
+		(_fromCalldata ? "calldata" : "memory") +
+		"_dynamic_" +
+		_type.identifier();
+
+	return m_functionCollector->createFunction(functionName, [&] {
+		if (auto arrayType = dynamic_cast<ArrayType const*>(&_type))
+		{
+			solAssert(!arrayType->isDynamicallySized(), "");
+			solAssert(!_fromCalldata, "");
+			//solAssert(_padToWords, "");
+
+			return Whiskers(R"(
+				function <functionName>(addr) -> value {
+					value := mload(addr)
+				}
+			)")
+			("functionName", functionName)
+			.render();
+		}
+
+		unsigned numBytes = _type.calldataEncodedSize(false);
```

Comment:
Not needed.

Reply:
To explain: this is really just about calldata, not about memory.

Reply:
Or, perhaps this is needed, but we should discriminate between memory and calldata.

Reply:
the original code did this in both cases

Reply:
Please only call `_type.calldataEncodedSize(false)` if you know that the type can be stored in calldata. Otherwise it might throw an exception.